File: backend/services/currency.py
"""Normalização de códigos monetários recebidos de fontes de mercado."""

from __future__ import annotations


_CURRENCY_ALIASES = {"R$": "BRL", "US$": "USD"}


def normalize_currency_code(currency: str | None, fallback: str = "BRL") -> str:
    """Retorna um código ISO 4217, incluindo marcadores históricos do COTAHIST."""
    normalized = (currency or "").strip().upper()
    resolved = _CURRENCY_ALIASES.get(normalized, normalized)
    return resolved if len(resolved) == 3 and resolved.isalpha() else fallback

File: frontend/lib/currency.ts
const currencyAliases: Record<string, string> = {
  "R$": "BRL",
  "US$": "USD",
};

export function normalizeCurrencyCode(currency: string | null | undefined, fallback = "BRL") {
  const normalized = (currency ?? "").trim().toUpperCase();
  const resolved = currencyAliases[normalized] ?? normalized;
  return /^[A-Z]{3}$/.test(resolved) ? resolved : fallback;
}
